agent/stats/engine.go

```diff
@@ -163,77 +202,87 @@ func (engine *DockerStatsEngine) removeAll() {
-		return
-	}
-}
 
-// addExistingContainers lists existing containers and adds them to the engine.
-func (engine *DockerStatsEngine) addExistingContainers() error {
-	listContainersResponse := engine.client.ListContainers(false, ecsengine.ListContainersTimeout)
-	if listContainersResponse.Error != nil {
-		return listContainersResponse.Error
-	}
-
-	for _, containerID := range listContainersResponse.DockerIDs {
-		engine.addContainer(containerID)
+	for task := range engine.tasksToHealthCheckContainers {
+		delete(engine.tasksToContainers, task)
 	}
-
-	return nil
 }
 
-// addContainer adds a container to the map of containers being watched.
-// It also starts the periodic usage data collection for the container.
-func (engine *DockerStatsEngine) addContainer(dockerID string) {
-	engine.lock.Lock()
-	defer engine.lock.Unlock()
-
+// addContainerUnsafe adds a container to the map of containers being watched.
+func (engine *DockerStatsEngine) addContainerUnsafe(dockerID string) (*StatsContainer, error) {
 	// Make sure that this container belongs to a task and that the task
 	// is not terminal.
 	task, err := engine.resolver.ResolveTask(dockerID)
 	if err != nil {
-		seelog.Debugf("Could not map container to task, ignoring, err: %v, id: %s", err, dockerID)
-		return
+		return nil, errors.Wrapf(err, "could not map container to task, ignoring container: %s", dockerID)
 	}
 
 	if len(task.Arn) == 0 || len(task.Family) == 0 {
-		seelog.Debugf("Task has invalid fields, id: %s", dockerID)
-		return
+		return nil, errors.Errorf("stats add container: invalid task fields, arn: %s, familiy: %s", task.Arn, task.Family)
 	}
 
 	if task.GetKnownStatus().Terminal() {
-		seelog.Debugf("Task is terminal, ignoring, id: %s", dockerID)
-		return
+		return nil, errors.Errorf("stats add container: task is terminal, ignoring container: %s, task: %s", dockerID, task.Arn)
 	}
 
+	seelog.Debugf("Adding container to stats watch list, id: %s, task: %s", dockerID, task.Arn)
+	statsContainer := newStatsContainer(dockerID, engine.client, engine.resolver)
+	engine.tasksToDefinitions[task.Arn] = &taskDefinition{family: task.Family, version: task.Version}
+
+	statsContainerTracked := false
+	if !engine.disableMetrics {
+		// Adding container to the map for collecting stats
+		statsContainerTracked = engine.addToStatsContainerMapUnsafe(task.Arn, dockerID, statsContainer, engine.containerMetricsMapUnsafe)
+	}
+
+	if dockerContainer, err := engine.resolver.ResolveContainer(dockerID); err != nil {
+		seelog.Debugf("Could not map container ID to container, container: %s, err: %s", dockerID, err)
+	} else if dockerContainer.Container.HealthStatusShouldBeReported() {
+		// Track the container health status
+		engine.addToStatsContainerMapUnsafe(task.Arn, dockerID, statsContainer, engine.healthCheckContainerMapUnsafe)
+		seelog.Debugf("Adding container to stats health check watch list, id: %s, task: %s", dockerID, task.Arn)
+	}
+
+	if !statsContainerTracked {
```

Comment:
this still doesn't read right. if stats container is not tracked, we should start tracking it. However, that's not what this variable is conveying. Can you rename this to be `watchStatsContainer`?